### ISSUE TITLE
recompute_grad computes gradient incorrectly when the same tensor is passed in multiple argument positions

### ISSUE THREAD

**Issue by davisyoshida** [tensorflow/tensorflow#36999]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Linux 5.5.3-arch1-1
- TensorFlow installed from (source or
binary): binary
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de 2.1.0
- Python version: 3.7.5
- CUDA/cuDNN version: 10.2/7
- GPU model and memory: 1070 Ti, 8GB

**Describe the current behavior**
Passing the same tensor as two arguments to a function decorated with `recompute_grad` leads to incorrect gradient computation.

**Describe the expected behavior**
The gradient should be computed properly. The `tf.custom_gradient` code uses `experimental_ref`s to deduplicate variables, and doing the same for input tensors will resolve this issue.

**Standalone code to reproduce the issue** 
```
import tensorflow as tf

@tf.recompute_grad
def broken_add(a, b):·
    return a + b·

x = tf.ones(3, dtype=tf.float32)
with tf.GradientTape() as g:
    g.watch(x)
    z = tf.reduce_sum(broken_add(x, x)) 
print(g.gradient(z, x).numpy())
```
This outputs `[4. 4. 4.]` instead of the correct value `[2., 2., 2.]`.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
The following is used to deduplicate variables in `tf.custom_gradient`, and adding the same for input tensors should resolve the issue.
https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/ops/custom_gradient.py#L410-L414

**Comment by amahendrakar** [comment]
@davisyoshida,
Was able to reproduce the issue with [TF 2.1](https://colab.sandbox.google.com/gist/amahendrakar/6565a6eda1a453d494c974a51dd07142/36999.ipynb), but it seems to be fixed in [TF-nightly](https://colab.sandbox.google.com/gist/amahendrakar/9295e2162c43a37ca288e1faa8923d3d/copy-of-36999.ipynb). Please check the attached gists. Thanks!

**Comment by davisyoshida** [comment]
Great! Sorry to waste your time.

**Comment by amahendrakar** [comment]
No worries @davisyoshida, happy to help.

Closing this issue as it resolved.
